File: modules/expo-cadenza-midi/ios/CadenzaMidiModule.swift
import CoreMIDI
import ExpoModulesCore

// Local Expo module bridging iOS CoreMIDI input to JavaScript.
//
// - emits "onMessage" per MIDI message with { deviceId, bytes, timestamp } where
//   timestamp is milliseconds since an arbitrary epoch (monotonic).
// - emits "onDeviceChange" with { type: "added" | "removed", device } on hot-plug.
// - exposes getDevices() returning the current input device list.
//
// SysEx aggregation: CoreMIDI may split SysEx across packets. This module
// accumulates bytes from F0..F7 per source endpoint before emitting a single
// message.

public class CadenzaMidiModule: Module {
  private var midiClient: MIDIClientRef = 0
  private var inputPort: MIDIPortRef = 0
  private var connectedSources: Set<MIDIEndpointRef> = []
  private var sysexBuffers: [MIDIEndpointRef: [UInt8]] = [:]

  public func definition() -> ModuleDefinition {
    Name("CadenzaMidi")

    Events("onMessage", "onDeviceChange")

    OnCreate {
      self.setupMidi()
    }

    OnDestroy {
      self.teardownMidi()
    }

    Function("getDevices") { () -> [[String: Any]] in
      return self.enumerateDevices()
    }
  }

  // MARK: - setup / teardown

  private func setupMidi() {
    let notifyBlock: MIDINotifyBlock = { [weak self] notificationPointer in
      guard let self = self else { return }
      let notification = notificationPointer.pointee
      if notification.messageID == .msgSetupChanged {
        DispatchQueue.main.async { self.handleSetupChange() }
      }
    }

    let clientName = "Cadenza" as CFString
    var status = MIDIClientCreateWithBlock(clientName, &midiClient, notifyBlock)
    guard status == noErr else { return }

    let portName = "CadenzaInput" as CFString
    let readBlock: MIDIReadBlock = { [weak self] packetListPtr, srcConnRefCon in
      guard let self = self else { return }
      guard let refCon = srcConnRefCon else { return }
      let endpoint = MIDIEndpointRef(UInt32(bitPattern: Int32(Int(bitPattern: refCon) & 0xFFFFFFFF)))
      self.handlePacketList(packetListPtr.pointee, endpoint: endpoint)
    }

    status = MIDIInputPortCreateWithBlock(midiClient, portName, &inputPort, readBlock)
    guard status == noErr else { return }

    connectAllSources()
  }

  private func teardownMidi() {
    if inputPort != 0 { MIDIPortDispose(inputPort); inputPort = 0 }
    if midiClient != 0 { MIDIClientDispose(midiClient); midiClient = 0 }
    connectedSources.removeAll()
    sysexBuffers.removeAll()
  }

  // MARK: - device enumeration

  private func enumerateDevices() -> [[String: Any]] {
    var out: [[String: Any]] = []
    let count = MIDIGetNumberOfSources()
    for i in 0..<count {
      let source = MIDIGetSource(i)
      out.append(deviceInfoDict(for: source))
    }
    return out
  }

  private func deviceInfoDict(for endpoint: MIDIEndpointRef) -> [String: Any] {
    return [
      "id": deviceId(for: endpoint),
      "name": deviceName(for: endpoint),
      "transport": "unknown"
    ]
  }

  private func deviceId(for endpoint: MIDIEndpointRef) -> String {
    var uniqueId: Int32 = 0
    MIDIObjectGetIntegerProperty(endpoint, kMIDIPropertyUniqueID, &uniqueId)
    return String(uniqueId)
  }

  private func deviceName(for endpoint: MIDIEndpointRef) -> String {
    var name: Unmanaged<CFString>?
    let status = MIDIObjectGetStringProperty(endpoint, kMIDIPropertyDisplayName, &name)
    if status == noErr, let cf = name?.takeRetainedValue() {
      return cf as String
    }
    return "Unknown MIDI"
  }

  // MARK: - source connection lifecycle

  private func connectAllSources() {
    let count = MIDIGetNumberOfSources()
    for i in 0..<count {
      let source = MIDIGetSource(i)
      if !connectedSources.contains(source) {
        let refCon = UnsafeMutableRawPointer(bitPattern: Int(source))
        if MIDIPortConnectSource(inputPort, source, refCon) == noErr {
          connectedSources.insert(source)
          sendEvent("onDeviceChange", [
            "type": "added",
            "device": deviceInfoDict(for: source)
          ])
        }
      }
    }
  }

  private func handleSetupChange() {
    let currentSources = Set((0..<MIDIGetNumberOfSources()).map { MIDIGetSource($0) })
    // Removed sources: disconnect + emit
    for source in connectedSources.subtracting(currentSources) {
      let info = deviceInfoDict(for: source)
      MIDIPortDisconnectSource(inputPort, source)
      connectedSources.remove(source)
      sysexBuffers.removeValue(forKey: source)
      sendEvent("onDeviceChange", ["type": "removed", "device": info])
    }
    // New sources: connect + emit
    connectAllSources()
  }

  // MARK: - message dispatch

  private func handlePacketList(_ packetList: MIDIPacketList, endpoint: MIDIEndpointRef) {
    var packet = packetList.packet
    for _ in 0..<packetList.numPackets {
      let length = Int(packet.length)
      let bytes: [UInt8] = withUnsafeBytes(of: packet.data) { rawPtr in
        Array(rawPtr.prefix(length))
      }
      dispatchBytes(bytes, endpoint: endpoint, timestamp: packet.timeStamp)
      packet = MIDIPacketNext(&packet).pointee
    }
  }

  private func dispatchBytes(_ bytes: [UInt8], endpoint: MIDIEndpointRef, timestamp: UInt64) {
    var buffer = sysexBuffers[endpoint] ?? []
    var i = 0

    while i < bytes.count {
      let byte = bytes[i]

      if !buffer.isEmpty {
        buffer.append(byte)
        if byte == 0xF7 {
          emit(bytes: buffer, endpoint: endpoint, timestamp: timestamp)
          buffer = []
        }
        i += 1
        continue
      }

      if byte == 0xF0 {
        buffer = [0xF0]
        i += 1
        continue
      }

      let length = CadenzaMidiModule.shortMessageLength(forStatus: byte)
      let end = min(i + length, bytes.count)
      emit(bytes: Array(bytes[i..<end]), endpoint: endpoint, timestamp: timestamp)
      i = end
    }

    if buffer.isEmpty {
      sysexBuffers.removeValue(forKey: endpoint)
    } else {
      sysexBuffers[endpoint] = buffer
    }
  }

  private static func shortMessageLength(forStatus status: UInt8) -> Int {
    let high = status & 0xF0
    switch high {
    case 0x80, 0x90, 0xA0, 0xB0, 0xE0: return 3
    case 0xC0, 0xD0: return 2
    case 0xF0:
      switch status {
      case 0xF1, 0xF3: return 2
      case 0xF2: return 3
      case 0xF6, 0xF8, 0xFA, 0xFB, 0xFC, 0xFE, 0xFF: return 1
      default: return 1
      }
    default: return 1
    }
  }

  private func emit(bytes: [UInt8], endpoint: MIDIEndpointRef, timestamp: UInt64) {
    // MIDITimeStamp is mach_absolute_time; convert to milliseconds.
    let ms = Double(timestamp) / 1_000_000.0
    sendEvent("onMessage", [
      "deviceId": deviceId(for: endpoint),
      "bytes": bytes.map { Int($0) },
      "timestamp": ms
    ])
  }
}

File: modules/expo-cadenza-midi/android/src/main/java/expo/modules/cadenzamidi/CadenzaMidiModule.kt
package expo.modules.cadenzamidi

import android.content.Context
import android.media.midi.MidiDevice
import android.media.midi.MidiDeviceInfo
import android.media.midi.MidiManager
import android.media.midi.MidiOutputPort
import android.media.midi.MidiReceiver
import android.os.Handler
import android.os.Looper
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Local Expo module bridging Android MidiManager input to JavaScript.
 *
 * - emits "onMessage" per MIDI message with { deviceId, bytes, timestamp } where
 *   timestamp is milliseconds since an arbitrary epoch (monotonic).
 * - emits "onDeviceChange" with { type: "added" | "removed", device } on hot-plug.
 * - exposes getDevices() returning the current input device list.
 *
 * SysEx aggregation: MidiManager packets can split SysEx mid-stream. This module
 * accumulates bytes from F0..F7 per source device before emitting a single message.
 */
class CadenzaMidiModule : Module() {
  private var midiManager: MidiManager? = null
  private val openDevices: MutableMap<Int, MidiDevice> = mutableMapOf()
  private val openPorts: MutableMap<Int, MutableList<MidiOutputPort>> = mutableMapOf()
  private val sysexBuffers: MutableMap<Int, MutableList<Byte>> = mutableMapOf()
  private val mainHandler = Handler(Looper.getMainLooper())

  private val deviceCallback = object : MidiManager.DeviceCallback() {
    override fun onDeviceAdded(info: MidiDeviceInfo) {
      openDevice(info)
    }

    override fun onDeviceRemoved(info: MidiDeviceInfo) {
      closeDevice(info)
    }
  }

  override fun definition() = ModuleDefinition {
    Name("CadenzaMidi")

    Events("onMessage", "onDeviceChange")

    OnCreate {
      val context = appContext.reactContext ?: return@OnCreate
      val mgr = context.getSystemService(Context.MIDI_SERVICE) as? MidiManager ?: return@OnCreate
      midiManager = mgr
      for (info in mgr.devices) {
        openDevice(info)
      }
      mgr.registerDeviceCallback(deviceCallback, mainHandler)
    }

    OnDestroy {
      midiManager?.unregisterDeviceCallback(deviceCallback)
      for ((_, ports) in openPorts) {
        for (port in ports) {
          try { port.close() } catch (_: Throwable) {}
        }
      }
      openPorts.clear()
      for ((_, device) in openDevices) {
        try { device.close() } catch (_: Throwable) {}
      }
      openDevices.clear()
      sysexBuffers.clear()
      midiManager = null
    }

    Function("getDevices") {
      return@Function midiManager?.devices?.map { deviceInfoToMap(it) } ?: emptyList<Map<String, Any?>>()
    }
  }

  private fun deviceInfoToMap(info: MidiDeviceInfo): Map<String, Any?> {
    val name = info.properties.getString(MidiDeviceInfo.PROPERTY_NAME)
      ?: info.properties.getString(MidiDeviceInfo.PROPERTY_PRODUCT)
      ?: "Unknown MIDI"
    val transport = when (info.type) {
      MidiDeviceInfo.TYPE_USB -> "usb"
      MidiDeviceInfo.TYPE_BLUETOOTH -> "bluetooth"
      MidiDeviceInfo.TYPE_VIRTUAL -> "virtual"
      else -> "unknown"
    }
    return mapOf(
      "id" to info.id.toString(),
      "name" to name,
      "transport" to transport,
    )
  }

  private fun openDevice(info: MidiDeviceInfo) {
    val mgr = midiManager ?: return
    if (openDevices.containsKey(info.id)) return

    mgr.openDevice(info, { device ->
      if (device == null) return@openDevice
      val id = info.id
      openDevices[id] = device

      val ports = mutableListOf<MidiOutputPort>()
      for (portIndex in 0 until info.outputPortCount) {
        val port = device.openOutputPort(portIndex) ?: continue
        val receiver = object : MidiReceiver() {
          override fun onSend(msg: ByteArray?, offset: Int, count: Int, timestampNs: Long) {
            if (msg == null) return
            val bytes = msg.copyOfRange(offset, offset + count)
            dispatchBytes(bytes, id, timestampNs)
          }
        }
        port.connect(receiver)
        ports.add(port)
      }
      openPorts[id] = ports

      sendEvent("onDeviceChange", mapOf(
        "type" to "added",
        "device" to deviceInfoToMap(info),
      ))
    }, mainHandler)
  }

  private fun closeDevice(info: MidiDeviceInfo) {
    val id = info.id
    val device = openDevices.remove(id) ?: return
    openPorts.remove(id)?.forEach { try { it.close() } catch (_: Throwable) {} }
    sysexBuffers.remove(id)
    try { device.close() } catch (_: Throwable) {}
    sendEvent("onDeviceChange", mapOf(
      "type" to "removed",
      "device" to deviceInfoToMap(info),
    ))
  }

  private fun dispatchBytes(bytes: ByteArray, deviceId: Int, timestampNs: Long) {
    val buffer = sysexBuffers[deviceId] ?: mutableListOf()

    var i = 0
    while (i < bytes.size) {
      val b = bytes[i].toInt() and 0xFF

      if (buffer.isNotEmpty()) {
        buffer.add(bytes[i])
        if (b == 0xF7) {
          emit(buffer.toByteArray(), deviceId, timestampNs)
          buffer.clear()
        }
        i++
        continue
      }

      if (b == 0xF0) {
        buffer.add(bytes[i])
        i++
        continue
      }

      val length = shortMessageLength(b)
      val end = minOf(i + length, bytes.size)
      emit(bytes.copyOfRange(i, end), deviceId, timestampNs)
      i = end
    }

    if (buffer.isEmpty()) {
      sysexBuffers.remove(deviceId)
    } else {
      sysexBuffers[deviceId] = buffer
    }
  }

  private fun shortMessageLength(status: Int): Int {
    return when (status and 0xF0) {
      0x80, 0x90, 0xA0, 0xB0, 0xE0 -> 3
      0xC0, 0xD0 -> 2
      0xF0 -> when (status) {
        0xF1, 0xF3 -> 2
        0xF2 -> 3
        else -> 1
      }
      else -> 1
    }
  }

  private fun emit(bytes: ByteArray, deviceId: Int, timestampNs: Long) {
    val ms = timestampNs.toDouble() / 1_000_000.0
    sendEvent("onMessage", mapOf(
      "deviceId" to deviceId.toString(),
      "bytes" to bytes.map { it.toInt() and 0xFF },
      "timestamp" to ms,
    ))
  }
}
